File: src/clipwright/recap_config.py
"""Per-project recap preferences.

Lives at `<project>/.clipwright/recap-config.json`. Carries the knobs
the user can dial in once and have apply across every video they
generate from that project — most importantly the target duration
(so Claude doesn't compress a 3-min story into a 60s ceiling) and an
outro spec (so every video in the project ends the same way).

Why a separate file (not part of project.json):
  - Lets the v2 schema stay minimal and stable.
  - Mirrors the pattern of `.clipwright/claude-permissions.json`,
    `claude-model.json`, `claude-timeout.json` — runtime knobs are
    project-scoped sidecars; manifest stays the source of truth for
    the deliverable.
  - Easy for the user to nuke and start over without touching the
    project manifest.

The agent prompt reads this on every turn and injects a "Project
preferences" section, so changes take effect immediately without a
restart.
"""
from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from pathlib import Path

CONFIG_REL_PATH = Path(".clipwright") / "recap-config.json"

# Sensible defaults — every recap project starts with 1:30 target
# duration and a 3-second outro. The user wanted a fixed default for
# both rather than the "0 = unset" sentinel we had before, so even an
# untouched project gives Claude an explicit target and the prompt
# injection ALWAYS runs.
DEFAULT_TARGET_DURATION = 90       # 1 min 30 sec
DEFAULT_OUTRO_DURATION = 3.0       # seconds
# When the user hasn't customized the outro description, we synthesize
# one referencing the project title in a cyberpunk visual idiom — see
# `effective_outro_description()` below.


@dataclass
class OutroSpec:
    """How the user wants every video in the project to end.

    `description` is plain natural-language — the agent reads it and
    composes a matching final segment. `duration_seconds` is the
    target outro length so the agent doesn't blow past it.

    An empty `description` is replaced at prompt-injection time with a
    cyberpunk-themed fallback that surfaces the project title — see
    `effective_outro_description()`.
    """

    description: str = ""
    duration_seconds: float = DEFAULT_OUTRO_DURATION


@dataclass
class PersonaDraft:
    """The desktop persona builder's field values.

    Pure UI state: `RecapConfig.persona` is the prose the agent prompt
    quotes, and these are the six blocks the builder composed it from
    (identity, tone, structural rules, vocabulary, pacing, and the
    forbidden failure mode). We persist them so reopening the persona
    panel resumes the build instead of stranding the user in free-text
    mode with prose they can no longer edit block-by-block. Nothing on
    the Python side reads these — they never reach the prompt.
    """

    role: str = ""
    voice: str = ""
    moves: str = ""
    vocabulary: str = ""
    pacing: str = ""
    avoid: str = ""


@dataclass
class RecapConfig:
    """User-tunable knobs that flow into the agent prompt.

    Defaults are deliberately non-zero/non-empty for the duration
    fields: every project should get a target duration and an outro
    duration even before the user opens settings. The agent prompt
    injection runs whenever any field deviates from a *pristine empty*
    state — which is always, given these defaults.
    """

    target_duration_seconds: int = DEFAULT_TARGET_DURATION
    narration_style: str = ""
    additional_notes: str = ""
    # Who Claude should BE when writing for this project — e.g. "an
    # expert manhwa scriptwriter who specializes in high-retention
    # hooks and dramatic pacing". Distinct from `narration_style`,
    # which describes the *voice actor* (timbre, delivery); the
    # persona describes the *writer* (expertise, editorial instincts).
    # Empty = no persona section in the prompt, so existing projects
    # keep behaving exactly as before.
    persona: str = ""
    # Builder state behind `persona` — see PersonaDraft. UI-only.
    persona_draft: PersonaDraft = field(default_factory=PersonaDraft)
    outro: OutroSpec = field(default_factory=OutroSpec)

    def to_dict(self) -> dict:
        return asdict(self)

    @classmethod
    def from_dict(cls, d: dict) -> RecapConfig:
        """Tolerant parse — unknown fields ignored, missing fields
        defaulted. Outro is upgraded from string-only legacy shape if
        we ever see it (defensive — we ship the structured shape from
        day one, but future-proof the loader)."""
        outro_raw = d.get("outro") or {}
        if isinstance(outro_raw, str):
            outro = OutroSpec(description=outro_raw)
        elif isinstance(outro_raw, dict):
            outro = OutroSpec(
                description=str(outro_raw.get("description", "")),
                duration_seconds=float(outro_raw.get("duration_seconds", 5.0)),
            )
        else:
            outro = OutroSpec()
        return cls(
            target_duration_seconds=int(d.get("target_duration_seconds", 0)),
            narration_style=str(d.get("narration_style", "")),
            additional_notes=str(d.get("additional_notes", "")),
            persona=str(d.get("persona", "")),
            persona_draft=_parse_persona_draft(d.get("persona_draft")),
            outro=outro,
        )

    def has_overrides(self) -> bool:
        """True when at least one field carries actionable content.

        With the new defaults (target=90s, outro=3s), this is
        effectively always True for a constructed `RecapConfig` —
        the prompt section is always emitted, which is what the
        user wanted ("video duration should always default to 1:30").
        Kept as a method so future fields can opt out of injection
        when truly unset.

        `persona` is deliberately NOT counted here: it renders as its
        own `## Persona` section with its own emptiness gate, so a
        persona-only config must not resurrect an otherwise-empty
        "Project preferences" block.
        """
        return (
            self.target_duration_seconds > 0
            or bool(self.narration_style.strip())
            or bool(self.additional_notes.strip())
            or bool(self.outro.description.strip())
            or self.outro.duration_seconds > 0
        )

    def effective_outro_description(self, project_title: str) -> str:
        """Resolve the outro description for prompt injection.

        When the user has typed nothing, fall back to a cyberpunk-
        themed description featuring the project title — that's the
        product's house style. The user can override anytime via the
        Project Settings dialog; an explicit description short-circuits
        this fallback.
        """
        if self.outro.description.strip():
            return self.outro.description.strip()
        title = project_title.strip() or "Clipwright"
        return (
            f"Cyberpunk-themed outro card. Project title \"{title}\" appears "
            "center-screen in a neon-accented sans-serif over a deep blue/violet "
            "background with subtle scanline/glitch. Voiceover (short, 1 line): "
            "tease the next chapter and prompt a follow."
        )


def _parse_persona_draft(raw: object) -> PersonaDraft:
    """Tolerant parse — anything that isn't a dict of strings yields an
    empty draft. This is UI state written by the desktop; a malformed
    value must never break loading the rest of the config."""
    if not isinstance(raw, dict):
        return PersonaDraft()
    return PersonaDraft(
        role=str(raw.get("role", "") or ""),
        voice=str(raw.get("voice", "") or ""),
        moves=str(raw.get("moves", "") or ""),
        vocabulary=str(raw.get("vocabulary", "") or ""),
        pacing=str(raw.get("pacing", "") or ""),
        avoid=str(raw.get("avoid", "") or ""),
    )


def config_path(project_dir: Path) -> Path:
    return Path(project_dir) / CONFIG_REL_PATH


def load_recap_config(project_dir: Path) -> RecapConfig:
    """Read the per-project recap config. Returns defaults when
    missing or malformed — never raises (this is a soft preferences
    file, not a manifest)."""
    path = config_path(Path(project_dir))
    if not path.exists():
        return RecapConfig()
    try:
        return RecapConfig.from_dict(json.loads(path.read_text()))
    except (json.JSONDecodeError, OSError, TypeError, ValueError):
        return RecapConfig()


def save_recap_config(project_dir: Path, cfg: RecapConfig) -> Path:
    """Atomic write — temp file + rename so a half-write doesn't
    leave the file corrupted on disk."""
    path = config_path(Path(project_dir))
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(".json.tmp")
    tmp.write_text(json.dumps(cfg.to_dict(), indent=2) + "\n")
    tmp.replace(path)
    return path

File: desktop/src-tauri/src/recap_config.rs
//! Per-project recap preferences — typed pass-through to the Python side.
//!
//! The config lives at `<project>/.clipwright/recap-config.json` and
//! carries the user's video-length target, narration style, and outro
//! spec. The Python module
//! `src/clipwright/recap_config.py` owns the schema + I/O; here we
//! just read/write the file directly so the desktop doesn't pay a
//! subprocess round-trip on every save.
//!
//! Why direct file I/O instead of shelling to the CLI: the config is
//! plain JSON with a stable shape and no validation that requires the
//! Python runtime. Keeping the parsing local makes the settings
//! dialog feel instant.

use std::path::PathBuf;

use serde::{Deserialize, Serialize};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum RecapConfigError {
    #[error("io: {0}")]
    Io(#[from] std::io::Error),
    #[error("malformed recap-config.json: {0}")]
    Parse(#[from] serde_json::Error),
}

impl Serialize for RecapConfigError {
    fn serialize<S: serde::Serializer>(&self, ser: S) -> Result<S::Ok, S::Error> {
        ser.serialize_str(&self.to_string())
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct OutroSpec {
    #[serde(default)]
    pub description: String,
    #[serde(default = "default_outro_duration")]
    pub duration_seconds: f64,
}

impl Default for OutroSpec {
    fn default() -> Self {
        Self {
            description: String::new(),
            duration_seconds: default_outro_duration(),
        }
    }
}

fn default_outro_duration() -> f64 {
    // Matches `DEFAULT_OUTRO_DURATION` in
    // `src/clipwright/recap_config.py`. If either side changes the
    // default, update both so a fresh project loads the same numbers
    // whether we go through the Python loader or the Rust one.
    3.0
}

fn default_target_duration() -> u64 {
    // Matches `DEFAULT_TARGET_DURATION` in
    // `src/clipwright/recap_config.py`.
    90
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct RecapConfig {
    /// Defaults to 90s (1:30). Override via the settings dialog when
    /// a project needs longer or shorter videos.
    #[serde(default = "default_target_duration")]
    pub target_duration_seconds: u64,
    #[serde(default)]
    pub narration_style: String,
    #[serde(default)]
    pub additional_notes: String,
    /// Who Claude should BE when writing for this project ("an expert
    /// manhwa scriptwriter who specializes in high-retention hooks…").
    /// Empty = no persona section in the agent prompt. Mirrors
    /// `RecapConfig.persona` in `src/clipwright/recap_config.py`.
    #[serde(default)]
    pub persona: String,
    #[serde(default)]
    pub outro: OutroSpec,
}

impl Default for RecapConfig {
    fn default() -> Self {
        Self {
            target_duration_seconds: default_target_duration(),
            narration_style: String::new(),
            additional_notes: String::new(),
            persona: String::new(),
            outro: OutroSpec::default(),
        }
    }
}

fn config_path(project_dir: &str) -> PathBuf {
    PathBuf::from(project_dir)
        .join(".clipwright")
        .join("recap-config.json")
}

/// Read the per-project recap config. Returns `RecapConfig::default()`
/// when the file is missing or malformed — soft-fail by design so the
/// settings dialog always opens to *something* the user can edit.
#[tauri::command]
pub async fn get_recap_config(project_dir: String) -> Result<RecapConfig, RecapConfigError> {
    let path = config_path(&project_dir);
    if !path.exists() {
        return Ok(RecapConfig::default());
    }
    let raw = std::fs::read_to_string(&path)?;
    // Soft-fail on parse errors too: a corrupted file shouldn't trap
    // the user into a broken dialog. They can re-save and we'll
    // overwrite with a clean copy.
    match serde_json::from_str::<RecapConfig>(&raw) {
        Ok(cfg) => Ok(cfg),
        Err(_) => Ok(RecapConfig::default()),
    }
}

#[tauri::command]
pub async fn set_recap_config(
    project_dir: String,
    config: RecapConfig,
) -> Result<(), RecapConfigError> {
    let path = config_path(&project_dir);
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    // Atomic write via temp + rename so a half-written file doesn't
    // corrupt the on-disk state.
    let tmp = path.with_extension("json.tmp");
    let serialized = serde_json::to_string_pretty(&config)?;
    std::fs::write(&tmp, serialized.as_bytes())?;
    std::fs::rename(&tmp, &path)?;
    Ok(())
}
